### DISCUSSION TITLE
undefined reference to `evp_cipher_fetch',EVP_MD_get_size while upgrading open ssl from 1.0.2 to 3.0.8

### DISCUSSION THREAD

**Discussion by pekumar** [openssl/openssl#22564] [Q&A] [answered]
I am able to build the image of openssl 3.0.8 version and when i try to use the openssl image to build kerberos with crytpo as openssl im getting undefined reference to `evp_cipher_fetch',EVP_MD_get_size.

Here is my docker commands that are used for installing openssl 3.0.8 which is running successfully

```
FROM debian:bullseye

RUN mkdir -p /var/installation-files/openssl3


RUN set -ex 
RUN apt-get update -y 
RUN apt-get -y install  curl perl build-essential libssl-dev gcc tar 

COPY ./enable_fips.sh /var/installation-files
RUN chmod +x /var/installation-files/enable_fips.sh
RUN /var/installation-files/enable_fips.sh

```
enable_fips.sh contains following code:
```
#!/bin/sh

set -e
cd /var/installation-files/

curl --output openssl-3.tar.gz https://www.openssl.org/source/openssl-3.0.8.tar.gz
tar -xf openssl-3.tar.gz
cd openssl-3.0.8

chmod +x config
./config
make
make test
make install

ln -s /usr/local/lib64/libssl.so.3 /usr/lib/libssl.so.3
ln -s /usr/local/lib64/libcrypto.so.3 /usr/lib/libcrypto.so.3
```
now using this as base image im trying to install kerberos  and here is my docker file for the same
```
FROM newopenssl:latest
# Setting up Environment Variable
ENV PATH=/usr/local/bin:$PATH \
    LANG=C.UTF-8 \
    DEBIAN_FRONTEND=noninteractive
ENV KRB5_KDC_PROFILE=/etc/krb5kdc/kdc.conf
# Setting up variable for Kerberos version
ARG KERBEROS_VERSION=1.18

# Download and Install Openssl and OpenSSL FIPS Component
RUN set -ex \
    && apt-get update \
    && apt-get -y install  perl build-essential bison flex libssl-dev xinetd sudo  wget curl\
    && mkdir -p /usr/local/src/ \
    && cd /usr/local/src/ \
    && curl -O https://kerberos.org/dist/krb5/${KERBEROS_VERSION}/krb5-${KERBEROS_VERSION}.tar.gz \
    && mkdir -p /var/lib/krb5kdc \
    && mkdir -p /etc/krb5kdc \
    && tar -xvf krb5-${KERBEROS_VERSION}.tar.gz \
    && cd krb5-${KERBEROS_VERSION}/src \
    && ./configure --with-crypto-impl=openssl --with-prng-alg=os --localstatedir=/var/lib/ \    # iam getting error in this line
    && make \
    && make install \
    && apt-get remove -y build-essential bison flex mailutils-common \
    && apt-get remove -y --purge mysql\* \
    && apt-get autoclean \
    && apt-get clean
-->
Please help me out to resolve this issue , Here is the error screens shot
<img width="999" alt="Screenshot 2023-10-31 at 2 35 08 PM" src="https://github.com/openssl/openssl/assets/142476796/9bda8cc7-8569-411e-820a-b11ca2759c84">

**Comment by pekumar** [comment]
error screen shot
<img width="999" alt="Screenshot 2023-10-31 at 2 35 08 PM" src="https://github.com/openssl/openssl/assets/142476796/2e5217d5-8faf-4458-9188-323a67eb8271">

**Comment by paulidale** [comment]
Are you linking against the 3.0.8 OpenSSL libraries?

**Reply by pekumar** [reply]
yes I am linking the 3.0.8 libraries to use it as crypto for kerberos installation 
here is the line where I am trying to use openssl 
&& ./configure --with-crypto-impl=openssl --with-prng-alg=os --localstatedir=/var/lib/ \

**Reply by pekumar** [reply]
@paulidale  can you please help me on it, I mentioned all the commands that I am using in the description of the issue , Please let me know if any of the step is wrong or missing.

**Reply by pekumar** [reply]
@paulidale @alex can anyone please help me on this

**Reply by t8m** [reply]
You need to have libcrypto.so and libssl.so symlinks pointing to 3.0 libraries as well. Otherwise the linker links to the 1.0 or 1.1 versions which do not have these symbols.

**Reply by pekumar** [reply]
How I can do that, any command suggestion please

**Reply by pekumar** [reply]
@t8m , i used these two commands to link 
ln -s /usr/local/lib64/libssl.so.3 /usr/lib/libssl.so.3
ln -s /usr/local/lib64/libcrypto.so.3 /usr/lib/libcrypto.so.3

please correct me if this is wrong

**Reply by pekumar** [reply]
@t8m ,can you please let me know the command required to be added or modified in the above mentioned docker commands to make libcrypto.so and libssl.so symlinks pointing to 3.0 libraries

**Reply by t8m** [reply]
This is OK but you probably also need
```
ln -s /usr/local/lib64/libssl.so.3 /usr/lib/libssl.so
ln -s /usr/local/lib64/libcrypto.so.3 /usr/lib/libcrypto.so
```

**Reply by pekumar** [reply]
so now anyway im using them, still I am getting the error

**Reply by t8m** [reply]
That means the krb5 library build is still not finding the right libcrypto.so. Try to find it on the image and replace it with the symlink.

**Reply by t8m** [reply]
Or, if it is possible, you can try to change the configure command line options for the krb5 library to explicitly point it at the openssl build in /usr/local/lib64

**Reply by pekumar** [reply]
how I can do that in kerberos docker file , I mean in which line I can modify this

**Reply by t8m** [reply]
Unfortunately I cannot help you with that.

**Reply by levitte** [reply]
As I pointed out below, that's a matter of touching environment variables for the `configure` call.

**Reply by pekumar** [reply]
ya actually , its bit confusing for me where I need to exactly modify and link the libraries properly to krb5 config, Hence I am trying to get some more help

**Comment by levitte** [comment]
1. you really shouldn't have to `ln` the libraries into `/usr/lib`.  Instead you can use `LDFLAGS` to set the extra path to look into, like this:

   `LDFLAGS="-L/usr/local/lib64" ./configure ...`

2. I fear that version of krb5 isn't adapted for OpenSSL 3...  if you're working on adapting it for OpenSSL 3 (that's quite a jump from 1.0.2), that's of course all right, but if not, be aware that you're likely to run into quite a lot of errors (I just tried a build myself, and yeah...)

**Reply by pekumar** [reply]
then what would be the best version to upgrade

**Reply by levitte** [reply]
Considering kerberos.org hasn't had an update since 2013, I'm not sure how to answer that.

**Reply by pekumar** [reply]
kerberos has the updated version available which is 1.21

**Reply by levitte** [reply]
Ah, wikipedia has a more encouraging answer: https://en.wikipedia.org/wiki/Kerberos_(protocol)
There's a version 1.21, which was released this year.

**Reply by levitte** [reply]
I suggest you try that version

**Reply by pekumar** [reply]
yup , I have an idea on this but my major problem is on how I can link the openssl 3.0 libraries to krb5 libraries so that I can come out of these errors. 
Yup I will modify the version of kerberos as well , but suggest me the way to link the libraries of openssl to krb5

**Reply by levitte** [reply]
Like I said already:

```
LDFLAGS="-L/usr/local/lib64" ./configure --with-crypto-impl=openssl --with-prng-alg=os --localstatedir=/var/lib/
```

(provided /usr/local/lib64 is where the OpenSSL libraries were installed, otherwise you'll have to adapt)

**Reply by levitte** [reply]
BTW, I can confirm that krb5 1.21 builds cleanly (apart from a few innocuous warnings) with OpenSSL 3.

**Reply by pekumar** [reply]
oh thank you , let me also try the same in my machine.

**Comment by levitte** [comment, marked as answer]
My earlier recommendation for building against a custom OpenSSL installation was a bit flawed, you need to specify the inclusion directory as well.  Like this:

```
LDFLAGS="-L/usr/local/lib64" CPPFLAGS="-I/usr/local/include" ./configure --with-crypto-impl=openssl --with-prng-alg=os --localstatedir=/var/lib/

**Reply by pekumar** [reply]
ya sure , I will try this in my machine

**Reply by pekumar** [reply]
Awesome it worked for me , thank you so much

**Reply by pekumar** [reply]
@levitte , to make the above openssl 3.0.8 version as FIPS compliant do I need to add anything in the bash script or by default it comes with FIPS?

**Reply by pekumar** [reply]
@levitte , can you please help me on this

**Reply by levitte** [reply]
To produce a FIPS provider from the current source, you must add `enable-fips` in your build.

For FIPS *compliance*, you have to build it separately and exactly as the current validation Security Policy tells you to do, see https://www.openssl.org/docs/fips.html and follow the link to the validation certificate.

**Reply by pekumar** [reply]
is this way is correct @levitte for making openssl FIPS compliant 

created a new sh file 
#!/bin/sh

cat << 'EOF' > /etc/ssl/openssl.cnf
config_diagnostics = 1
openssl_conf = openssl_init

.include /usr/local/ssl/fipsmodule.cnf

[openssl_init]
providers = provider_sect

[provider_sect]
fips = fips_sect
base = base_sect

[base_sect]
activate = 1

EOF

docker file modifications are here :

FROM debian:bullseye
RUN mkdir -p /var/installation-files/openssl3
RUN set -ex 
RUN apt-get update -y 
RUN apt-get -y install  curl perl build-essential libssl-dev gcc tar 

COPY ./enable_fips.sh /var/installation-files
RUN chmod +x /var/installation-files/enable_fips.sh
RUN /var/installation-files/enable_fips.sh \
    && ln -f -s /etc/ssl/openssl.cnf /usr/local/ssl/openssl.cnf \
    && apt-get remove -y build-essential curl \
    && apt-get autoclean \
    && apt-get clean

# Adding and Executing Script to configure OpenSSL for FIPS mode
ADD fips_config.sh /
RUN chmod +x fips_config.sh && ./fips_config.sh

@levitte can you please validate once and let me know if this is correct ? Thank you.
